compiler/src/iree/compiler/GlobalOptimization/test/raise_special_ops.mlir

```diff
@@ -538,3 +538,125 @@ util.func public @conv_nchw_extf_both(%arg0 : tensor<1x5x10x10xf16>,
 //  CHECK-SAME:     %[[ARG1:.+]]: tensor<5x5x3x3xf16>
 //       CHECK:   %[[RESULT:.+]] = linalg.conv_2d_nchw_fchw {{.*}} ins(%[[ARG0]], %[[ARG1]]
 //       CHECK:   util.return %[[RESULT]]
+
+// -----
+
+util.func public @matmul_extsi(%arg0 : tensor<10x20xi32>,
+                               %arg1 : tensor<20x40xi16>) -> tensor<10x40xi32> {
+  %0 = tensor.empty() : tensor<20x40xi32>
+  %1 = linalg.generic {
+      indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0, d1)>],
+      iterator_types = ["parallel", "parallel"]}
+      ins(%arg1 : tensor<20x40xi16>) outs(%0 : tensor<20x40xi32>) {
+    ^bb0(%b0 : i16, %b1 : i32):
+      %e = arith.extsi %b0 : i16 to i32
+      linalg.yield %e : i32
+  } -> tensor<20x40xi32>
+  %2 = tensor.empty() : tensor<10x40xi32>
+  %3 = arith.constant 0 : i32
+  %4 = linalg.fill ins(%3 : i32) outs(%2 : tensor<10x40xi32>) -> tensor<10x40xi32>
+  %5 = linalg.matmul ins(%arg0, %1 : tensor<10x20xi32>, tensor<20x40xi32>)
```

Comment:
If this was a `linalg.matmul_unsigned` this patterns would be incorrect because it would change the extension semantics. This is why we need to handle each op one by one for integer extends because different named ops have different extension semantics, unlike float ops which all imply `extf`

Reply:
I'm not quite sure how this doesnt work, printing out the body of the `matmul_unsigned` shows that it is extended as a signed integer

```mlir
module {
  util.func public @matmul_extsi(%arg0: tensor<10x20xi32>, %arg1: tensor<20x40xi16>) -> tensor<10x40xi32> {
    %c0_i32 = arith.constant 0 : i32
    %0 = tensor.empty() : tensor<10x40xi32>
    %1 = linalg.fill ins(%c0_i32 : i32) outs(%0 : tensor<10x40xi32>) -> tensor<10x40xi32>
    %2 = linalg.matmul_unsigned ins(%arg0, %arg1 : tensor<10x20xi32>, tensor<20x40xi16>) outs(%1 : tensor<10x40xi32>) -> tensor<10x40xi32>
    util.return %2 : tensor<10x40xi32>
  }
}

// Generic form
#map = affine_map<(d0, d1, d2) -> (d0, d2)>
#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
"builtin.module"() ({
  "util.func"() <{function_type = (tensor<10x20xi32>, tensor<20x40xi16>) -> tensor<10x40xi32>, sym_name = "matmul_extsi", sym_visibility = "public", tied_operands = [-1 : index]}> ({
  ^bb0(%arg0: tensor<10x20xi32>, %arg1: tensor<20x40xi16>):
    %0 = "arith.constant"() <{value = 0 : i32}> : () -> i32
    %1 = "tensor.empty"() : () -> tensor<10x40xi32>
    %2 = "linalg.fill"(%0, %1) <{operandSegmentSizes = array<i32: 1, 1>}> ({
    ^bb0(%arg5: i32, %arg6: i32):
      "linalg.yield"(%arg5) : (i32) -> ()
    }) : (i32, tensor<10x40xi32>) -> tensor<10x40xi32>
    %3 = "linalg.matmul_unsigned"(%arg0, %arg1, %2) <{operandSegmentSizes = array<i32: 2, 1>}> ({
    ^bb0(%arg2: i32, %arg3: i16, %arg4: i32):
      %4 = "arith.extsi"(%arg3) : (i16) -> i32
      %5 = "arith.muli"(%arg2, %4) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
      %6 = "arith.addi"(%arg4, %5) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
      "linalg.yield"(%6) : (i32) -> ()
    }) {linalg.memoized_indexing_maps = [#map, #map1, #map2]} : (tensor<10x20xi32>, tensor<20x40xi16>, tensor<10x40xi32>) -> tensor<10x40xi32>
    "util.return"(%3) : (tensor<10x40xi32>) -> ()
  }) : () -> ()
}) : () -> ()

```

Reply:
If that's the case then that's a bug. If I take the IR you just posted and run `linalg-generalize-named-ops` I get an extui
```
#map = affine_map<(d0, d1) -> ()>
#map1 = affine_map<(d0, d1) -> (d0, d1)>
#map2 = affine_map<(d0, d1, d2) -> (d0, d2)>
#map3 = affine_map<(d0, d1, d2) -> (d2, d1)>
#map4 = affine_map<(d0, d1, d2) -> (d0, d1)>
module {
  util.func public @matmul_extsi(%arg0: tensor<10x20xi32>, %arg1: tensor<20x40xi16>) -> tensor<10x40xi32> {
    %c0_i32 = arith.constant 0 : i32
    %0 = tensor.empty() : tensor<10x40xi32>
    %1 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel"]} ins(%c0_i32 : i32) outs(%0 : tensor<10x40xi32>) {
    ^bb0(%in: i32, %out: i32):
      linalg.yield %in : i32
    } -> tensor<10x40xi32>
    %2 = linalg.generic {indexing_maps = [#map2, #map3, #map4], iterator_types = ["parallel", "parallel", "reduction"]} ins(%arg0, %arg1 : tensor<10x20xi32>, tensor<20x40xi16>) outs(%1 : tensor<10x40xi32>) {
    ^bb0(%in: i32, %in_0: i16, %out: i32):
      %3 = arith.extui %in_0 : i16 to i32
      %4 = arith.muli %in, %3 : i32
      %5 = arith.addi %out, %4 : i32
      linalg.yield %5 : i32
    } -> tensor<10x40xi32>
    util.return %2 : tensor<10x40xi32>
  }
}

```

Reply:
yeah, seems like a bug. `linalg-generalize-named-op` is producing different output than `mlir-print-op-generic`

Reply:
ugh yeah I think the hidden region on linalg named ops is probably broken then. In all other contexts `matmul_unsigned` is assumed to mean `extui`. Looking at the list of linalg operations: https://mlir.llvm.org/docs/Dialects/Linalg/#operations

it looks like only matmul_unsigned has these unsigned extension semantics, so we can just add a special case in the pattern to fail if trying to fuse with a `matmul_unsigned`.

Reply:
Okay, looked at this further. `linalg-generalize-named-ops` does print `extsi`, but I wasn't running against the mlir I posted here. I started with a `generic` (signed cast) -> `matmul_unsigned` and used the modified pass from this pr. This manually places the `extsi` inside of the `matmul_unsigned`'s region, but it doesn't get printed out. I'm not sure if changing regions like this is allowed/good.

Maybe it makes sense to only fold extsi with signed variants

Reply:
> I'm not sure if changing regions like this is allowed/good.

Agree :) although I see it more as a function of named ops being over engineered. It's probably because we're trying to use interfaces instead of just using the named op builders, but linalg is awkward...